aamp/json_util.py

```diff
@@ -0,0 +1,139 @@
+            'objects': {_try_name(None, idx, k, plist._crc32): self.encode_pobject(pobj) for idx, (k, pobj) in enumerate(plist.objects.items())},
+        }
+
+    def encode_pobject(self, obj: aamp.ParameterObject) -> dict:
+        return {
+            'params': {_try_name(None, idx, k, obj._crc32): self.encode_param(param) for idx, (k, param) in enumerate(obj.params.items())}
+        }
+
+    def encode_param(self, param) -> dict:
+        type_name = type(param).__name__
+        if getattr(parameters, type_name, None) is None:
+            return param
+        return {
+            '_type': type(param).__name__,
+            'value': asdict(param) if is_dataclass(param) else param
+        }
+
+
+def aamp_to_json(data: Union[bytes, aamp.Reader, dict], pretty: bool = False) -> str:
+    if isinstance(data, bytes):
+        data = aamp.Reader(data)
+    if isinstance(data, aamp.Reader):
+        data = data.parse()
+    return json.dumps(data, cls=AampJSONEncoder, ensure_ascii=False, indent=4 if pretty else None)
+
+
+class AampJSONDecoder(json.JSONDecoder):
+    def __init__(self, *args, **kwargs):
+        json.JSONDecoder.__init__(self, object_hook=self.object_hook, *args, **kwargs)
+
+    def object_hook(self, o):
+        try:
+            iter(o)
+        except:
+            return o
+        obj_type = type(o).__name__
+        if getattr(parameters, obj_type, None) is not None:
+            return o
+        if getattr(aamp, obj_type, None) is not None:
+            return o
+        if 'version' in o:
+            return self._to_pio(o)
+        else:
+            return o
+
+    def _to_pobj(self, o) -> aamp.ParameterObject:
+        if isinstance(o, aamp.ParameterObject):
+            return o
+        pobj = aamp.ParameterObject()
+        if o['params']:
+            for param, val in o['params'].items():
+                if param.isnumeric():
+                    pobj.params[int(param)] = self._to_param(val)
+                else:
+                    pobj.set_param(param, self._to_param(val))
+        return pobj
+
+    def _to_plist(self, o) -> aamp.ParameterList:
+        plist = aamp.ParameterList()
+        if isinstance(o, aamp.ParameterList):
+            return o
+        if o['lists']:
+            for name, content in o['lists'].items():
+                if name.isnumeric():
+                    plist.lists[int(name)] = self._to_plist(content)
+                else:
+                    plist.set_list(name, self._to_plist(content))
+        if o['objects']:
+            for name, content in o['objects'].items():
+                if content['params']:
+                    if name.isnumeric():
+                        plist.objects[int(name)] = self._to_pobj(content)
+                    else:
+                        plist.set_object(name, self._to_pobj(content))
+        return plist
+
+    def _to_pio(self, o) -> aamp.ParameterIO:
+        pio = aamp.ParameterIO(o['type'], o['version'])
+        if o['lists']:
```

Comment:
This is partially a duplicate of self._to_plist. Maybe consider moving the shared logic into a separate internal function to avoid duplicating code